org.eclipse.xtext.xtext.generator/src/org/eclipse/xtext/xtext/generator/validation/ValidatorFragment2.xtend

```diff
@@ -123,38 +127,48 @@ class ValidatorFragment2 extends AbstractInheritingFragment {
 			}
 		''').writeTo(projectConfig.runtime.src)
 	}
-	
+
 	protected def generateGenValidator() {
 		// take the non-abstract class signature for the src-gen class in case of !generateStub
-		//  as validators of sub languages refer to 'superGrammar.validatorClass',
-		//  see 'getGenValidatorSuperClass(...)'
-		val genClass = if (isGenerateStub) grammar.abstractValidatorClass else grammar.validatorClass
+		// as validators of sub languages refer to 'superGrammar.validatorClass',
+		// see 'getGenValidatorSuperClass(...)'
+		val genClass = if(isGenerateStub) grammar.abstractValidatorClass else grammar.validatorClass
 
 		val javaFile = fileAccessFactory.createGeneratedJavaFile(genClass)
-		
 		javaFile.content = '''
 			�IF !composedChecks.empty�
-			@�ComposedChecks�(validators = {�FOR validator: composedChecks SEPARATOR ", "��validator.typeRef�.class�ENDFOR�})
+				@�ComposedChecks�(validators = {�FOR validator: composedChecks SEPARATOR ", "��validator.typeRef�.class�ENDFOR�})
 			�ENDIF�
 			public �IF isGenerateStub�abstract �ENDIF�class �genClass.simpleName� extends �grammar.genValidatorSuperClass� {
 				
 				@Override
 				protected �List�<�EPackage�> getEPackages() {
 					�List�<�EPackage�> result = new �ArrayList�<�EPackage�>(�IF inheritImplementation && grammar.nonTerminalsSuperGrammar !== null�super.getEPackages()�ENDIF�);
-					�FOR e: generatedPackagesToValidate�
+					�FOR e : generatedPackagesToValidate�
 						result.add(�e.generatedEPackageName�.eINSTANCE);
 					�ENDFOR�
-					�FOR e: registryPackagesToValidate�
+					�FOR e : registryPackagesToValidate�
 						result.add(EPackage.Registry.INSTANCE.getEPackage("�e.nsURI�"));
-				 	�ENDFOR�
+					�ENDFOR�
 					return result;
 				}
-				
+				�generateValidationToDeprecateRules�
 			}
 		'''
 		javaFile.writeTo(projectConfig.runtime.srcGen)
 	}
-	
+
+	protected def StringConcatenationClient generateValidationToDeprecateRules() '''
+		�FOR deprecatedRule : deprecatedRulesFromGrammar�
+			�val elementType = new TypeReference(deprecatedRule.type.classifier as EClass, grammar.eResource.resourceSet)�
+			
+			@�Check�
+			public void checkDeprecated�elementType.simpleName�(�elementType� element) {
```

Comment:
would be nice to opt this out e.g. when i want to have the deprecation on the grammar level only

Reply:
Sure.